Initial commit of a generation tool, readme, and generator (#1)
DeterminateSystems/hydra-github-jobsets-generator: README.md

# Generate declarative jobsets for Hydra for a GitHub Project

> This tool uses Hydra's _declarative jobsets_ to automatically create Hydra jobsets for your GitHub project.

Reviewer: Do we want to link to Hydra's docs on declarative jobsets as well? https://github.com/NixOS/hydra/blob/master/doc/manual/src/plugins/declarative-projects.md or some recent build on h.n.o?

> Configuring declarative jobsets use several points of indirection, and this repository's goal is to simplify and the setup process.

Reviewer: "simplify and the setup process"

Maybe you were aiming for:

```suggestion
Configuring declarative jobsets use several points of indirection, and this repository's goal is to simplify and streamline the setup process.
```

```json
{
    "enabled": 1,
    "hidden": false,
    "description": "GitHub Pull Request Jobset Generator",
    "nixexprinput": "generator",
    "nixexprpath": "jobset/generate.nix",
    "checkinterval": 300,
    "schedulingshares": 100,
    "enableemail": false,
    "emailoverride": "",
    "keepnr": 3,
    "inputs": {
        "src": {
            "type": "git",
            "value": "https://github.com/YOURORGNAME/YOURREPONAME.git main",
            "emailresponsible": false
        },
        "generator": {
            "type": "git",
            "value": "https://github.com/DeterminateSystems/hydra-github-jobsets-generator.git main",
            "emailresponsible": false
        },
        "nixpkgs": {
            "type": "git",
            "value": "git://github.com/NixOS/nixpkgs.git nixos-unstable-small",
            "emailresponsible": false
        },
        "pull_requests": {
            "type": "githubpulls",
            "value": "YOURORGNAME YOURREPONAME",
            "emailresponsible": false
        }
    }
}
```

Take care to replace both `YOURORGNAME` and `YOURREPONAME` with your GitHub organization and repo's names.

### Configuring your repository: inputs

> Then, create a file at `.hydra/inputs.json`, and include all of the Hydra inputs you want in addition to your project's code.

Reviewer: Right now, this reads to me like "include the Hydra inputs you want and *do* add your project". Maybe something like the following would be clearer:

```suggestion
Then, create a file at `.hydra/inputs.json`, and include all of the other Hydra inputs you want to utilize.
```

For example, if your project depends on Nixpkgs, write the following to `.hydra/inputs.json`:

```json
{
    "nixpkgs": {
        "type": "git",
        "value": "git://github.com/NixOS/nixpkgs.git nixos-unstable-small",
        "emailresponsible": false
    }
}
```

If your project has no other inputs, write `{}` to the file.

### Configuring your Hydra

> Create a project on your Hydra which uses declarative input of typee `git`, and the file pointing to the `.hydra/project.json`.

Reviewer: ```suggestion
Create a project on your Hydra which uses a declarative input of type `git` with the file pointing to `.hydra/project.json`.
```

If you use the [Terraform Hydra provider](https://registry.terraform.io/providers/DeterminateSystems/hydra/latest), it would look like this:

```
resource "hydra_project" "pr-example" {
  name         = "pr-example"
  display_name = "PR Example"
  description  = "An example of building a GitHub repository's PRs."
  owner        = "alice"
  enabled      = true
  visible      = true

  declarative {
      file = ".hydra/project.json"
      type = "git"
      value = "https://github.com/YOURORGNAME/YOURREPONAME.git main"
  }
}
```

### Recap

> When the `.jobsets` jobset evaluates and builds you should now see a jobset for every open PR. This jobset will re-evaluate every 300 seconds (5 minutes.)

Reviewer: ```suggestion
When the `.jobsets` jobset evaluates and builds you should now see a jobset for every open PR. This jobset will re-evaluate every 300 seconds (5 minutes).
```
